### PR TITLE
Add a method to validate the presence of liquid tag 

### PR DESCRIPTION
**Pull request by melashu**
In this PR I have done the following changes 

- On `spec/liquid_validations_spec.rb` line 1 change `require 'spec_helper'` to `require_relative 'spec_helper'`
- Added a method `validates_liquid_tag` to `liquid-validations.rb` file
- Added a test to verify the functionality of the `validates_liquid_tag` method
- Added how to user validates_liquid_tag` method to README file

### CODE REVIEW

File: spec/liquid_validations_spec.rb [hunk 1]
```diff
@@ -62,4 +66,101 @@ class Mixin < ActiveRecord::Base
...
+     @mixin.errors.full_messages.any? { |e| e == "content must not have more than max tags 2"}.must_equal true
+   end
+
+   it 'must be valid when using more complex tag' do
+     @mixin.content = "{% josh_is_awesome foobar, data-required='true' %}"
+     @mixin.valid?
+     @mixin.errors.full_messages.any? { |e| e == "You must supply {% josh_is_awesome %} in your content"}.must_equal false
+   end
+
+   it 'must be invalid when using more complex tag' do
```
Comment (espen) [inline spec/liquid_validations_spec.rb]: looks like this and some other tests are duplicated?

File: spec/liquid_validations_spec.rb [hunk 2]
```diff
@@ -62,4 +66,101 @@ class Mixin < ActiveRecord::Base
...
+     @mixin.errors.full_messages.any? { |e| e == "You must supply {% josh_is_awesome %} in your content"}.must_equal true
+   end
+
+   it 'must be invalid when using tag like empty content' do
+     @mixin.valid?
+     @mixin.errors.full_messages.any? { |e| e == "You must supply {% josh_is_awesome %} in your content"}.must_equal true
+   end
+
+   it 'must be invalid when tag is greater than max count' do
+     @mixin.content = '{% josh_is_awesome %} {% josh_is_awesome %} {% josh_is_'
```
Comment (espen) [inline spec/liquid_validations_spec.rb]: the third tag here is incomplete: {% josh_is_

also this test should not fail with the current content. So add a second test with correct tag, and keep this test but make sure it is valid.

File: spec/liquid_validations_spec.rb [hunk 3]
```diff
@@ -1,4 +1,4 @@
-require 'spec_helper'
+require_relative 'spec_helper'
```
Comment (espen) [inline spec/liquid_validations_spec.rb]: why did you change this?

**Review by espen — CHANGES_REQUESTED** [review]

File: spec/liquid_validations_spec.rb (hunk 3 above)
Comment (melashu) [reply spec/liquid_validations_spec.rb]: I changed this because it fails to run the test, it gives me a load error like below

image

File: spec/liquid_validations_spec.rb (hunk 2 above)
Comment (melashu) [reply spec/liquid_validations_spec.rb]: Yes with this test it should be valid I will add another test with correct tag

**Comment by espen** [comment]
Looks good. I will try it in my app.

File: lib/liquid-validations.rb [hunk 4]
```diff
@@ -42,6 +42,25 @@ def validates_presence_of_liquid_variable(*attr_names)
...
+    raise(ArgumentError, 'You must supply a tag or max to check for ') if configuration[:tag].blank? || configuration[:max].blank? 
+    validates_each attr_names, configuration do |record, attr_name, value|
+      value    = value.to_s
+      max      = configuration[:max]
+      tag   = configuration[:tag].to_s
+      tag_r = /{%\s+#{tag}\s+(.*?)%}/
+      if !(value =~ tag_r)
+        record.errors.add(:base, "You must supply {% #{tag} %} in your #{ friendly_attr_name(attr_name) }")
+      elsif max && (value.scan(tag_r).size > max)
+       record.errors.add(:base, "#{friendly_attr_name(attr_name)} must not have more than max tags #{max}")
```
Comment (espen) [inline lib/liquid-validations.rb]: include tag name in error message

Comment (espen) [reply lib/liquid-validations.rb]: not fixed

File: spec/liquid_validations_spec.rb [hunk 5]
```diff
@@ -62,4 +66,118 @@ class Mixin < ActiveRecord::Base
...
+
+    it 'must be invalid when using tag like empty content' do
+      @mixin.valid?
+      @mixin.errors.full_messages.any? { |e| e == "You must supply {% josh_is_awesome %} in your content"}.must_equal true
+    end
+
+    it 'must be valid when tag like {% josh_is_awesome %} {% josh_is_awesome %} {% josh_is_' do
+      @mixin.content = '{% josh_is_awesome %} {% josh_is_awesome %} {% josh_is_'
+      @mixin.valid?
+      @mixin.errors.full_messages.any? { |e| e == "content must not have more than max tags 2"}.must_equal false
```
Comment (espen) [inline spec/liquid_validations_spec.rb]: test must also be updated accordingly with tag name. And error message should be similar in wording to the other error.

**Review by espen — CHANGES_REQUESTED** [review]

File: spec/liquid_validations_spec.rb [hunk 6]
```diff
@@ -62,4 +66,118 @@ class Mixin < ActiveRecord::Base
...
+
+    it 'must be invalid when using tag like empty content' do
+      @mixin.valid?
+      @mixin.errors.full_messages.any? { |e| e == "You must supply {% josh_is_awesome %} in your content"}.must_equal true
+    end
+
+    it 'must be valid when tag like {% josh_is_awesome %} {% josh_is_awesome %} {% josh_is_' do
+      @mixin.content = '{% josh_is_awesome %} {% josh_is_awesome %} {% josh_is_'
+      @mixin.valid?
+      @mixin.errors.full_messages.any? { |e| e == "content must not have more than 2 {% josh_is_awesome %}"}.must_equal false
```
Comment (espen) [inline spec/liquid_validations_spec.rb]: this should be valid. there are only two tags of {% josh_is_awesome %} here. the '{% josh_is_' is not a complete tag.

**Comment by espen** [comment]
Presence does not seem to work. If I have content without the tags that specify presence: false I still get error message.

Example:

validates_liquid_tag :booking_form, tag: 'street', max: 1, presence: false

content:
{% name %}
{% phone %}
{% email %}
{% date_of_birth %}

error:
You must supply {% street %} in your booking form

This should not give error.

File: spec/liquid_validations_spec.rb [hunk 7]
```diff
@@ -62,4 +66,118 @@ class Mixin < ActiveRecord::Base
...
+      @mixin.errors.full_messages.any? { |e| e == "You must supply {% josh_is_awesome %} in your content"}.must_equal true
+    end
+
+    it 'must be invalid when using tag like {% josh_is_awesome }' do
+      @mixin.content = "{% josh_is_awesome }"
+      @mixin.valid?
+      @mixin.errors.full_messages.any? { |e| e == "You must supply {% josh_is_awesome %} in your content"}.must_equal true
+    end
+
+    it 'must be invalid when using tag like { josh_is_awesome %}' do
```
Comment (espen) [inline spec/liquid_validations_spec.rb]: duplicate of test on line 135?

**Comment by espen** [comment]
Also make sure it is valid for presence: false when the content is nil

**Comment by espen** [comment]
Looks good. This seems to work in my app now.

I have additional wishes. You can try to look at it but if it is too difficult then don't spend too much time on it.

1) Have `presence` argument accept a Proc. Example:
```validates_liquid_tag :booking_form, tag: 'phone', max: 1, presence: Proc.new {|user|user.verification_method == 'sms'}```

2) Allow `tag` to be an array. Example:
```validates_liquid_tag :booking_form, tag: ['name', 'email', 'phone'], max: 1, presence: false```

File: spec/liquid_validations_spec.rb (hunk 6 above)
Comment (espen) [reply spec/liquid_validations_spec.rb]: Can you check this?

**Comment by espen** [comment]
Also check the review comment.

**Comment by espen** [comment]
I was not able to use the Proc because it doesn't work like ActiveModel `:if` Proc. See https://api.rubyonrails.org/v7.0.7/classes/ActiveModel/Validations/ClassMethods.html#method-i-validates_each

I looked briefly into the Rails code but was unable to see how it passes the object into the Proc. For `:if` you can do this: `if: Proc.new { |user| user.signup_step > 2 }` when the model is `User` so it will pass the context of user object.

Maybe have a quick look at it if you are interested. But don't spend too much time on it.

**Review by espen — APPROVED** [review]
